### PR TITLE
[SPARK-31344][CORE] Polish implementation of barrier() and allGather()

### PR DESCRIPTION
**Pull request by Ngone51**
### What changes were proposed in this pull request?

1. Combine  `BarrierRequestToSync` and `AllGatherRequestToSync` into `RequestToSync`, which is distinguished by `RequestMethod` type.

2. Remove unnecessary Json serialization/deserialization

3. Clean up some codes to make runBarrier() and `BarrierCoordinator` more general

4. Remove unused imports.

### Why are the changes needed?

To make codes simpler for better maintain in the future.

### Does this PR introduce any user-facing change?

No.

### How was this patch tested?

This is pure code refactor, so should be covered by existed tests.

### CODE REVIEW

**Comment by Ngone51** [comment]
This PR tried to address comment https://github.com/apache/spark/pull/27395#pullrequestreview-357681136 , but please note that we can not simply build `barrier()` on top of `allGather("")`, because we need a way do distinguish these two different calls to prevent following illegal case:

```scala
rdd.barrier().mapPartitions { iter =>
   val context = BarrierTaskContext.get()
   val partitionId = context.partitionId
   if (partitionId == 0) {
     context.barrier()
   } else {
     context.allGather(partitionId.toString)
   }
  iter
 }
```

File: core/src/main/scala/org/apache/spark/BarrierCoordinator.scala
Comment (Ngone51) [inline core/src/main/scala/org/apache/spark/BarrierCoordinator.scala:252]: TODO: remove redundant empty line.

File: core/src/main/scala/org/apache/spark/BarrierTaskContext.scala
Comment (sarthfrey) [inline core/src/main/scala/org/apache/spark/BarrierTaskContext.scala:173]: nit: extra line

**Review by sarthfrey — APPROVED** [review]
Thanks for cleaning this up @Ngone51, writing back the messages individually rather than parsing JSON + keeping a single case class seems to have removed a lot of bloat :)

**Comment by Ngone51** [comment]
Ping @mengxr @jiangxb1987

**Comment by jiangxb1987** [comment]
LGTM, I'll wait for another couple days before merge this PR.

**Comment by jiangxb1987** [comment]
retest this please

**Comment by jiangxb1987** [comment]
retest this please

**Comment by jiangxb1987** [comment]
retest this please

**Comment by jiangxb1987** [comment]
retest this please

**Comment by jiangxb1987** [comment]
retest this please

**Comment by jiangxb1987** [comment]
Thanks, merged to master!

**Comment by Ngone51** [comment]
thanks all!

**Pull request closed without merging**
